src/main/flight/position.c

```diff
@@ -22,205 +22,202 @@
-                zeroedAltitudeCm = gpsAltCm - gpsAltOffsetCm; // now that we have a GPS offset value, we can use it to zero relativeAltitude
-            }
-        } else {
-            gpsTrust = 0.0f;
-            // TO DO - smoothly reduce GPS trust, rather than immediately dropping to zero for what could be only a very brief loss of 3D fix 
-        }
-        DEBUG_SET(DEBUG_ALTITUDE, 2, lrintf(zeroedAltitudeCm / 10.0f)); // Relative altitude above takeoff, to 0.1m, rolls over at 3,276.7m
-        
-        // Empirical mixing of GPS and Baro altitudes
-        if (useZeroedGpsAltitude && (positionConfig()->altitude_source == DEFAULT || positionConfig()->altitude_source == GPS_ONLY)) {
-            if (haveBaroAlt && positionConfig()->altitude_source == DEFAULT) {
-                // mix zeroed GPS with Baro altitude data, if Baro data exists if are in default altitude control mode 
-                const float absDifferenceM = fabsf(zeroedAltitudeCm - baroAltCm) / 100.0f * positionConfig()->altitude_prefer_baro / 100.0f;
-                if (absDifferenceM > 1.0f) { // when there is a large difference, favour Baro
-                    gpsTrust /=  absDifferenceM;
+        switch (positionConfig()->altitude_source) {
+            case ALTITUDE_SOURCE_DEFAULT:
+#if defined(USE_GPS) && defined(USE_BARO)
+                // Fuse GPS and baro for more precise altitude
+                if (isGpsAvailable && isBaroAvailable) {  // Complementary filter
+                    gpsAltLowpassCm = pt1FilterApply(&gpsFuseLpf, gpsAltCm - gpsAltGroundCm);
+                    baroAltHighpassCm = baroAltCm - pt1FilterApply(&baroFuseLpf, baroAltCm);
+                    altitudeCm = gpsAltLowpassCm + baroAltHighpassCm;
+                } else
+#endif
+                if (isGpsAvailable && !isBaroAvailable) {  // Fallback option if baro fails
+                    altitudeCm = gpsAltCm - gpsAltGroundCm;
+                } else if (!isGpsAvailable && isBaroAvailable) {  // Fallback option if GPS fails
+                    altitudeCm = baroAltCm;
                 }
-                zeroedAltitudeCm = zeroedAltitudeCm * gpsTrust + baroAltCm * (1.0f - gpsTrust);
-            }
-        } else if (haveBaroAlt && (positionConfig()->altitude_source == DEFAULT || positionConfig()->altitude_source == BARO_ONLY)) {
-            zeroedAltitudeCm = baroAltCm; // use Baro if no GPS data, or we want Baro only
+                break;
+            case ALTITUDE_SOURCE_GPS_ONLY:
+                altitudeCm = gpsAltCm - gpsAltGroundCm;
+                break;
+            case ALTITUDE_SOURCE_BARO_ONLY:
+                altitudeCm = baroAltCm;
+                break;
+            default:
+                break;
         }
     }
 
-    zeroedAltitudeCm = pt2FilterApply(&altitudeLpf, zeroedAltitudeCm);
-    // NOTE: this filter must receive 0 as its input, for the whole disarmed time, to ensure correct zeroed values on arming
-
     if (wasArmed) {
```

Comment:
`wasArmed == ARMING_FLAG(ARMED)` here and `wasArmed` is confusing

Reply:
I didn't change this because it is historic code. Is this supposed to be a bugfix?

Reply:
sorry, I'm just reading modified/related code. This is a problem with existing code, not your PR.